### ISSUE TITLE
Model evaluation

### ISSUE THREAD

**Issue by sanagno** [LAION-AI/Open-Assistant#1908]
We need a better evaluation pipeline to better quantify model performance and compare models with each other.
Some ideas include
- Evaluating on dataset for which we already have the chatGPT reference, e.g. HC3.
- Using a fine-tuned reward model.

**Comment by sanagno** [comment]
If we want to test knowledge/retrieval etc, there are some good evaluations on things like 'wizard of wikipedia' or 'wizard of the internet. 
If we want to test conversational skills there is 'blender skill talk' or 'convai2'. 
There are also things like 'Empathetic dialogue' to test empathy for instance

**Comment by pruksmhc** [comment]
Adding instructions evaluation dataset to this thread:
https://arxiv.org/pdf/2204.07705.pdf

**Comment by pruksmhc** [comment]
I'm interested to help with this, but have my hands tied for the next week or so.

**Comment by totuta** [comment]
Is this task being worked on right now?

**Comment by sanagno** [comment]
So as far as I know :)

**Comment by adendek** [comment]
Do you still need a contributor?

**Comment by sanagno** [comment]
yes, please, If anyone is interested I can assign you to it

**Comment by adendek** [comment]
I'll start working on this tomorrow, but I don't want to stop or delay anyone who wants to contribute as well.

**Comment by shahules786** [comment]
For using RM to evaluate LLM results I have pushed a framework to do this in a sister repo https://github.com/Open-Assistant/oasst-model-eval/tree/main/model_eval/scoring. I can replicate the same here if that's something we consider useful. @sanagno

**Comment by xrsrke** [comment]
@sanagno @adendek hey, i'm interested in help you guys with this. how can i help?

**Comment by adendek** [comment]
Can someone like @sanagno confirm if the code provided by @shahules786 is sufficient? You @ shahules786 can create PR which can be reviewed. 
If this happens, I will take care of a different task. 
___

However, my initial plan was as follows: 

- [ ] Clone the repo.
- [ ] Try to build and run the model on my own hardware or docker.
- [ ] Finish reading project documentation, and contribution page. 
- [ ] Find the location where the code should be located. 
- [ ] Implement some abstractions and expandable pipelines to ensure we can easily extend the set of evaluation metrics and methods. 
- [ ] Implement initial unit tests, since TDD rocks. 
- [ ] Review the paper provided by @pruksmhc to find an initial set of metrics to be used. 
- [ ] implement evaluation metrics and create work-in-progress PR to get quick feedback from the community. 

@xrsrke if you want to participate you can start on your own if you have any thoughts you can share them in this thread. Any help and suggestion will be appreciated. 

___

Let me know if this makes any sense. 

Adam

**Comment by xrsrke** [comment]
Thanks @adendek. I will start reading the paper tomorrow, follow @adendek 's initial plan and share my progress here!

**Comment by xrsrke** [comment]
After checking out the `Open-Assistant` and [`oasst-model-eval`](https://github.com/Open-Assistant/oasst-model-eval) repository (suggested by @shahules786), I found that they currently only have a sampling report and no evaluation on datasets.

So, my initial plan is to create evaluation scripts that can be extended to different datasets and metrics. @adendek, we should consider splitting the work to avoid duplication of effort.

**Comment by adendek** [comment]
@xrsrke great job! I updated the road map from my previous message. The point regarding the paper review can be considered done. 

Now, the highest priority is to understand the open assistant part. We need to find where the evaluation code should be placed, and how to get a model to validate it versus the tasks. We need a method or class that will implement the following interface: 


```python 
import torch
from torch.utils.data import Dataset

def evaluation_task(model: torch.nn.Module, dataset: Dataset) -> float:
  """ 
   Evaluates the model versus a given task represented as a PyTorch Dataset. 
  
  Parameters
  ----------
   model : torch.nn.Module
         the model to be evaluated. 
   dataset :  Dataset
         instance of a PyTorch dataset that contains the data for a given evaluation task.
  """

  implementation goes here.
```

Do we use numpy-like or google docstrings? 

___

> evaluation scripts that can be extended to different datasets and metrics

I'd rather say we need a little extendable framework, rather than a script. I prefer the OOP approach with proper interfaces and unit tests. There is also an issue with task orchestration. 

Do we need to create our solution or leverage an existing one like [airflow](https://airflow.apache.org/), [kedros](https://medium.com/quantumblack/introducing-kedro-the-open-source-library-for-production-ready-machine-learning-code-d1c6d26ce2cf), or [Luigi](https://luigi.readthedocs.io/en/stable/)? 
Where this evaluation pipeline should be evaluated? GitHub workflow? 

___

What do you think about this?

**Comment by sanagno** [comment]
Hey the code of @shahules786 seems to be taking care of the second part of the evaluation, that is great.

Regarding the first point, i.e evaluating on some known datasets your suggestions seems great. 

I think the most suitable location for the code is under `/model/model_eval`

**Comment by shahules786** [comment]
@adendek I can raise the PR for using RM to evaluate  RLHF fine-tuned model. FYI There is some refactoring going on with `reward-model` training #2071

**Comment by adendek** [comment]
@shahules786 sure do this. 

And I will take care of the rest of the framework. You should expect a WIP (work in progress) PR very soon, like Tuesday or Wednesday.

**Comment by xrsrke** [comment]
@adendek Something just came up with my work, and I won't be able to join anymore. Sorry for the inconvenience.

**Comment by shahules786** [comment]
@sanagno @adendek I'm waiting for the RM training team to release loadable model weights. This might take some days. I am anyway through with the rest of the code. As soon as the model is ready I'll raise the PR.

**Comment by andreaskoepf** [comment]
Reward model based evaluation is now available thanks @shahules786. 

@adendek, are you working on other benchmarks, e.g.  evaluating the instruction following capabilities?

**Comment by tju01** [comment]
I'm going to work on this now over at https://github.com/tju01/oasst-automatic-model-eval. I started with the OpenAI evals framework which also had its own issue #2348, but I'm going to extend that code to also cover over benchmarks now.

**Comment by andreaskoepf** [comment]
@tju has generated https://tju01.github.io/ilm-eval/
